### PR TITLE
Tweaks to query re-planning

### PR DESCRIPTION
**Pull request by systay**
- Renamed confusing variable names
- Make sure we compare the same stats
- Added more tests
- Tweaked the defaults to re-plan less often

### CODE REVIEW

**Comment by systay** [comment]
Related: #7784

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/PlanFingerprint.scala [hunk 1]
```diff
@@ -24,20 +24,21 @@ import org.neo4j.helpers.Clock
...
+      check(currentTxId != f.txId,
+        () => { fingerprint = Some(f.copy(creationTimeMillis = currentTimeMillis)) }) &&
       check(f.snapshot.diverges(f.snapshot.recompute(statistics), statsDivergenceThreshold),
-            { fingerprint = Some(f.copy(creationTimeMillis = currentTimeMillis, txId = currentTxId)) })
+        () => { fingerprint = Some(f.copy(creationTimeMillis = currentTimeMillis, txId = currentTxId)) })
     }
   }
 
-  private def check(test: => Boolean, ifFalse: => Unit ) = if (test) { true } else { ifFalse ; false }
+  private def check(test: => Boolean, ifFalse: () => Unit ) = if (test) { true } else { ifFalse ; false }
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/PlanFingerprint.scala:43]: maybe just me but I think it is easier to read if you call `ifFalse` explicitly,  i.e. `ifFalse()`

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/InstrumentedGraphStatistics.scala [hunk 2]
```diff
@@ -54,12 +56,14 @@ case class GraphStatisticsSnapshot(map: Map[StatisticsKey, Double] = Map.empty)
...
   //statistics that is bigger than the threshold
   def diverges(snapshot: GraphStatisticsSnapshot, minThreshold: Double): Boolean = {
-    assert(map.keySet == snapshot.map.keySet)
-    val v1 = map.values
-    val v2 = snapshot.map.values
+    assert(statsValues.keySet == snapshot.statsValues.keySet)
     //find the maximum relative difference (|e1 - e2| / max(e1, e2))
-    val relativeDiff = v1.zip(v2).map(e => Math.abs(e._1 - e._2) / Math.max(e._1, e._2)).max
-    relativeDiff > minThreshold
+    val divergedStats = (statsValues map {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/InstrumentedGraphStatistics.scala:61]: but it is so many lines :)

**Comment by pontusmelke** [comment]
Looks like reasonable changes, however we must be aware that this will change behaviour in a maintenance release and it will not be for the better for all workloads.

**Comment by davidegrohmann** [comment]
@pontusmelke this is actually fixing a bug, we didn't compare stats properly. The tweak of defaults is a nice addition to the PR.

**Comment by pontusmelke** [comment]
Oh I didn't realise, where is the actual bug fix?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/PlanFingerprint.scala (hunk 1 above)
Comment (jexp) [reply community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/PlanFingerprint.scala:43]: I think this whole complex statement is not easy to read. I'd prefer if those were individual statements perhaps with a clear variable name to hold the results, and potentially an early return.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/InstrumentedGraphStatistics.scala (hunk 2 above)
Comment (jexp) [reply community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/InstrumentedGraphStatistics.scala:61]: I like it :)

**Comment by jexp** [comment]
@pontusmelke comparing the two maps driven by keys not by `values1.zip(values2)`

**Comment by davidegrohmann** [comment]
@systay what's the status of this PR?

**Comment by systay** [comment]
@davidegrohmann It's cursed. The build keeps failing, but on unrelated problems such as maven central not responding in time and weird I/O failures. Will rebase to trigger a new build, and then hopefully it can go in.

**Comment by boggle** [comment]
+1

**Pull request merged**